core/src/main/java/de/danoeh/antennapod/core/sync/EpisodeActionFilter.java

```diff
@@ -0,0 +1,84 @@
+package de.danoeh.antennapod.core.sync;
+
+import android.util.Log;
+
+import androidx.collection.ArrayMap;
+import androidx.core.util.Pair;
+
+import java.util.List;
+import java.util.Map;
+
+import de.danoeh.antennapod.net.sync.model.EpisodeAction;
+
+public class EpisodeActionFilter {
+
+    public static final String TAG = "EpisodeActionFilter";
+
+    public Map<Pair<String, String>, EpisodeAction> getRemoteActionsOverridingLocalActions(
+            List<EpisodeAction> remoteActions,
+            List<EpisodeAction> queuedEpisodeActions) {
+        // make sure more recent local actions are not overwritten by older remote actions
+        Map<Pair<String, String>, EpisodeAction> remoteActionsThatOverrideLocalActions = new ArrayMap<>();
+        Map<Pair<String, String>, EpisodeAction> localMostRecentPlayActions =
+                createUniqueLocalMostRecentPlayActions(queuedEpisodeActions);
+        for (EpisodeAction remoteAction : remoteActions) {
+            Log.d(TAG, "Processing remoteAction: " + remoteAction.toString());
+            Pair<String, String> key = new Pair<>(remoteAction.getPodcast(), remoteAction.getEpisode());
+            switch (remoteAction.getAction()) {
+                case NEW:
+                    remoteActionsThatOverrideLocalActions.put(key, remoteAction);
+                    break;
+                case DOWNLOAD:
+                    break;
+                case PLAY:
+                    EpisodeAction localMostRecent = localMostRecentPlayActions.get(key);
+                    if (localActionHappenedAfterRemoteAction(remoteAction, localMostRecent)
+                    ) {
+                        break;
+                    }
+                    EpisodeAction mostRecentAction = remoteActionsThatOverrideLocalActions.get(key);
+                    if (
+                            mostRecentAction == null
+                                    || mostRecentAction.getTimestamp() == null
+                                    || (remoteAction.getTimestamp() != null
+                                    && mostRecentAction.getTimestamp().before(remoteAction.getTimestamp()))
+                    ) {
```

Comment:
Please reduce the used vertical space. Also, please use braces to make it more explicit what the operator precedence is (I must admit that I don't know from the top of my head how `||` and `&&` play together and I think many others don't know either).

Reply:
if you read carefully again you'll notice that the conjunction (&& operation) of the last two conditions is already grouped by braces. There is no need for any more braces.

I'll reduce vertical space.